Issue#2231: restore 'url' option for ImportControl

ivanov-alex opened the pull request.

mkordas commented:
@ivanov-alex 

1. Please make TeamCity happy.
2. Can you try to keep 100 characters in line even for tests? It's bit easier to read and review such code. I know we should enforce it some day.
3. `testUrlPositive()` - does this test depend on internet connection? If yes, we need to find some workaround. Our build must work without network.
4. `@Test(expected = AssertionError.class)` - that's risky and strange, can you explain? Tests should test production code, in production code no `AssertionError` is thrown.
5. `https://{WrongChars}.raw.githubusercontent.com/checkstyle/checkstyle/master/src/test/resources/com/puppycrawl/tools/checkstyle/imports/import-control_one.xml` - do we need so long URL to just put there some wrong chars?

ivanov-alex commented:
Done. For *#*3 I have replaced `https://` protocol with `file://`. It may take a while for me to make sure it works on Windows build as well due to different path building (I may need to trigger CI several times)

ivanov-alex commented:
It should be fine now.
the only thing I cannot understand is CI fail for not complete coverage on AbstractDeclarationCollector.java and SuppressWarningsHolder.java. Is it something caused by my changes?

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheck.java:
```diff
@@ -63,6 +64,11 @@
     public static final String MSG_DISALLOWED = "import.control.disallowed";
 
+    /**
+     * A key with message for exception.
```
Can you correct this Javadoc? It doesn't contain 'key'.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheckTest.java:
```diff
@@ -182,4 +182,45 @@ public void testWrongFormatURI() throws Exception {
+                + "InputImportControl.java"), expected);
+    }
+
+    @Test(expected = CheckstyleException.class)
```
Such assertion doesn't check the place where exception was thrown. We need to use try-catch here.

romani replied:
@ivanov-alex , we even do have Check in sevntu project that can allow forbid usage of "expected" in "@Test". In most cases usage of "expected" is laziness of developer - should be refactored with try-catch and recheck part of message to be sure that we catching expected Exception.

ivanov-alex replied:
In this particular case the best solution I see is:
```java
        try {
            verify(checkConfig, getPath("imports" + File.separator
                    + "InputImportControl.java"), expected);
        }
        catch (final CheckstyleException ex) {
            return;
        }
        fail();
```
Basically, it only narrows error scope to `verify` method, it doesn't guarantee that we have caught correct exception. 
I have two tests for two different situations. In both cases error is re-thrown with message similar to `cannot initialize module com.puppycrawl.tools.checkstyle.TreeWalker - Cannot set property 'url' to 'https://{WrongCharsInURL}' in module com.puppycrawl.tools.checkstyle.checks.imports.ImportControlCheck`.
This message doesn't contain anything from original error message and doesn't allow us to identify if we caught correct exception.

romani replied:
```
    verify(checkConfig, getPath("imports" + File.separator
                    + "InputImportControl.java"), expected);
        fail();
        }
        catch (final CheckstyleException ex) {
           assert(ex.getMessage().............)
        }
```

ivanov-alex replied:
Yes, this code is a bit better, but the main problem still the same: analysis of exception message doesn't makes sense, it has no context of original problem. I receive two identical messages for two different problems.

mkordas replied:
@ivanov-alex you can make assertion on `ex.getCause()`

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java:
```diff
@@ -93,6 +93,10 @@ protected static String getPath(String filename)
         return new File("src/test/resources/com/puppycrawl/tools/checkstyle/" + filename).getCanonicalPath();
     }
 
+    protected static String getURI(String filename) {
```
It's a bit confusing: `getURI` returning `String` used in `testUrl` test.

romani replied:
please make it `getUriString` , abbreviation is also need to be avoided.

mkordas commented:
@ivanov-alex you can make assertion on `getCause()`

ivanov-alex commented:
I was confused by the fact that for InvocationTargetException cause goes to `target` field.
Now got it resolved.

ivanov-alex commented:
Done.
Also added similar analysis of exception message for two tests with exceptions.
And dropped `testWrongFormatURI` test with `expected` (it appeared to test the same as `testUnknown`)

mkordas commented:
I think that `((InvocationTargetException) ex.getCause().getCause()).getTargetException().getMessage()` could be extracted to some `private static` helper method to decrease code duplication.Besides that changes look good for me.

romani commented:
merged as FF.

I am against moving such code to method, it is not that frequent/spread to carte about duplication.

mkordas commented:
@romani, exactly the same complicated sequence appears 4 times in one class, for me it was candidate for extraction.

romani commented:
ok, as private method in this class. Could be provided as next PR.
